Azure blob in flat folder structure

yogigrantz opened the pull request:
Added an option to output Azure blob in flat folder structure with custom filename prefix

Summary of the changes (Less than 80 chars)
 - Added internal static string to store the custom prefix
 - Added public overload to include the custom name parm
 - Modified existing nested folder creation to accept the flat folder option

Addresses #25885

previous discussion: https://github.com/dotnet/extensions/pull/3391

Commit:
Added an option to output Azure blob in flat folder structure with custom filename prefix

wtgodbe commented:
@JunTaoLuo who owns this package now, since it's moved from Extensions?

Pilchie commented:
@davidfowl do you know who knows about this package?

shirhatti commented:
Pavel created the package and no one has touched it since.

Pilchie commented:
Also tagging @javiercn.

davidfowl commented on src/Logging.AzureAppServices/src/BlobLoggerProvider.cs:
```diff
@@ -58,10 +58,20 @@ public BlobLoggerProvider(IOptionsMonitor<AzureBlobLoggerOptions> options)
             {
                 var key = eventGroup.Key;
-                var blobName = $"{_appName}/{key.Year}/{key.Month:00}/{key.Day:00}/{key.Hour:00}/{_fileName}";
+                //var blobName = $"{_appName}/{key.Year}/{key.Month:00}/{key.Day:00}/{key.Hour:00}/{_fileName}";
```
Why is this commented out?

yogigrantz replied:
That code was replaced with the logic from line 66 - 74 because it forces nested folder in Azure blob. Line 71 produces a flat folder structure, while line 74 produces the original nested folder structure where the code was commented out.

davidfowl replied:
Shouldn't it be removed then? Why did you leave the code in but commented out?

yogigrantz replied:
Yes you're right, it should be removed. I will remove it and commit my change

yogigrantz replied:
@davidfowl : I removed the commented out code, committed and pushed. This is my first time contributing, and I was not sure whether to leave commented out original code or to delete. But now I know to go ahead and delete when deemed not needed. Thank you.

yogigrantz replied:
@davidfowl : After removing the commented our code, it would not pass the checks. Could you please advise what might have gone wrong, could this be something on the CI side?

yogigrantz replied:
@davidfowl : The build is ready for your review. Could you please revisit this PR? Thank you!

Commit:
Remove the commented code - not needed anymore

wtgodbe commented:
We were having some issues with our selenium tests, they should be fixed now. This should pass on retry

wtgodbe commented:
/azp run

yogigrantz commented:
> Azure Pipelines successfully started running 2 pipeline(s).

still failing? hmmm.... could we try again please?

wtgodbe commented:
/azp run

wtgodbe commented:
@yogigrantz the CI failures are unrelated, we're having some issues w/ our template tests at the moment. Should hopefully be resolved soon.

yogigrantz commented:
> @yogigrantz the CI failures are unrelated, we're having some issues w/ our template tests at the moment. Should hopefully be resolved soon.

Thank you William!

wtgodbe commented:
/azp run

yogigrantz commented:
> /azp run

still failing ... I should have deleted that line in the first place  :(

Pilchie commented:
/azp run

Pilchie commented:
Let's give it another shot.

Pilchie commented:
@davidfowl or @javiercn - can one of you review this?

yogigrantz commented:
> Let's give it another shot.

All checks have passed now :) Thank you All!

yogigrantz commented:
@davidfowl @javiercn @Pilchie @wtgodbe @JunTaoLuo @shirhatti 
Could I please have this pull request approved before the CI automatically closes this issue. This feature is very important to our company to avoid workaround custom/duplicate Extension API. 

I can put sample Azure application in GitHub if that helps?. Please do let me know. Thank you so much, appreciate your time!

BrennanConroy commented:
@davidfowl Could you take a quick look at this and then we can bring it to API review?

davidfowl commented on src/Logging.AzureAppServices/src/AzureAppServicesLoggerFactoryExtensions.cs:
```diff
@@ -17,6 +17,22 @@ namespace Microsoft.Extensions.Logging
     /// </summary>
     public static class AzureAppServicesLoggerFactoryExtensions
     {
+        internal static string CustomPrefix;
```
This doesn't work because multiple independent calls in the same assembly will cause this to be overwritten.

yogigrantz replied:
This assembly is designed to be called only once when the IHostBuilder container is configured in Program.cs of a web application. Please refer to the next comment.

yogigrantz replied:
The concept of Logging Provider is that it is called once throughout the application lifetime. It's either with CustomPrefix or without. And the only place this value is set is at the entry point where the provider is injected to IHostBuilder ConfigureLogging. Nowehere else. Please re-review this change. Thank you. - cc: @BrennanConroy

davidfowl commented on src/Logging.AzureAppServices/src/AzureAppServicesLoggerFactoryExtensions.cs:
```diff
@@ -17,6 +17,22 @@ namespace Microsoft.Extensions.Logging
+        public static ILoggingBuilder AddAzureWebAppDiagnostics(this ILoggingBuilder builder, string customPrefix)
+        {
+            CustomPrefix = customPrefix;
+            var context = WebAppContext.Default;
```
The custom prefix should be passed into the constructor.

yogigrantz replied:
@davidfowl 
This is inside a static class, where there is no constructor. 

AddAzureWebAppDiagnostics is an overload extension of ILoggingBuilder, which is the "entry point" to this API. The web application program injects this API to the IHost container like this:

 ihost.ConfigureLogging(x =>
            {
                x.AddAzureWebAppDiagnostics(customPrefixName);
            });

as opposed to the original version, where the DI is coded like this:

 ihost.ConfigureLogging(x =>
            {
                x.AddAzureWebAppDiagnostics();
            });

Notice that all I did was adding that customPrefixName option and store its value in a static field where the provider checks whether to use the custom prefixed flat blob filename logic. I didn't want to touch the original method if possible, out of respect to the original programmer. Please let me know if there is a more preferable approach to implement this logic

yogigrantz replied:
This is a static class extension entry point to IHostBuilder ConfigureLogging Dependency Injection. It takes the responsibility of a constructor. I'm sure you are familiar with the use of Logging Provider Dependency Injection to IHostBuilder ConfigureLogging. Please revisit this code, I appreciate your time!

davidfowl replied:
We don't design anything like this, statics are pretty much forbidden, especially mutable ones like this. This needs to be a constructor argument to the instance.

davidfowl commented on src/Logging.AzureAppServices/src/BlobLoggerProvider.cs:
```diff
@@ -58,10 +58,19 @@ public BlobLoggerProvider(IOptionsMonitor<AzureBlobLoggerOptions> options)
-                var blobName = $"{_appName}/{key.Year}/{key.Month:00}/{key.Day:00}/{key.Hour:00}/{_fileName}";
+                string blobName;
+
+                if (!string.IsNullOrEmpty(AzureAppServicesLoggerFactoryExtensions.CustomPrefix))
```
This shouldn't be looked up via a static, but instead should be an instance.

yogigrantz replied:
This entire Azure Blob / File Log API is only called once in the beginning of the program run. I could create a new singleton class that contains this one string, but I believe that would be overkill. Unless that is the preferred approach? Please let me know. 
I have tested this solution in Azure and it works flawlessly both ways. The original way, where the blob is structured under appname/yyyy/mm/dd/hh/appInstanceId_blobname, and the modified way, where the blob is named {customprefix}{yyyy}(mm){dd} .

The only difference in setup is whether or not to pass the customPrefix name

ihost.ConfigureLogging(x =>
{
x.AddAzureWebAppDiagnostics(**customPrefixName**);
});

or

ihost.ConfigureLogging(x =>
{
x.AddAzureWebAppDiagnostics**()**;
});

Commit:
Changed approach

Changed approach - Removed mutable static object. Instead store the object (customPrefixFileName) in AzureBlobLoggerOptions and initialize it at DI entry point

yogigrantz reviewed (commented):
Removed the mutable object and store the customPrefixFilename string in the AzureBlobLoggerOptions

yogigrantz commented:
@wtgodbe  There are build issue on this module, error RS0016: (NETCORE_ENGINEERING_TELEMETRY=Build) Symbol 'AddAzureWebAppDiagnostics' is not part of the declared API. Could you please help me out? Thank you

wtgodbe commented:
Looks like we need to update the API baselines since this is adding new API - @dougbu could you remind me how to do that?

dougbu commented:
> @dougbu could you remind me how to do that?

I strongly recommend **not** trying to change the PublicAPI.Unshipped.txt files manually. The analyzer has fairly nice fixers that work only in Visual Studio. So,
1. Open a (preferably small) `*.slnf` file containing the project with API changes in VS
2. Build the solution
3. Click on one of the errors
4. Use the light bulb 💡 or right click to fix the error _globally_. I say _globally_ because it's a pain to click on every error and fix it individually
5. For errors at a lower level than added, changed, or removed surface area, you may need to either fix the API or suppress the error e.g. https://github.com/dotnet/aspnetcore/blob/master/src/Components/Components/src/EventCallbackFactoryBinderExtensions.cs#L35

dougbu commented:
Oh, it may be necessary to repeat the above in VS because the first round didn't get far enough to hit all of the `public` API issues

davidfowl commented on src/Logging.AzureAppServices/src/AzureAppServicesLoggerFactoryExtensions.cs:
```diff
@@ -21,15 +21,17 @@ public static class AzureAppServicesLoggerFactoryExtensions
-        public static ILoggingBuilder AddAzureWebAppDiagnostics(this ILoggingBuilder builder)
+        /// <param name="customPrefix"></param>
+        /// <returns></returns>
+        public static ILoggingBuilder AddAzureWebAppDiagnostics(this ILoggingBuilder builder, string customPrefix = null)
```
@Pilchie is this a binary breaking change?

Pilchie replied:
Absolutely.

Pilchie replied:
Overloads are binary breaking safe, optional parameters are not - think about how they work, the method always receives two arguments, it's just that if you don't specify it, the compiler burns in the default value at the call-site.

yogigrantz replied:
My bad, I removed the optional parm and went ahead and created an overload instead. But CI still shows this optional parm error, even though I renamed the parameter.
@Pilchie @davidfowl  
[SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters"

Commit:
Fixed binary breaking change

Removed the optional parameter, created overload to handle the extra parameter

Commit:
Renamed parm

Gihub repository did not take my change for removig the optional = null, so I am renaming the parm

yogigrantz commented:
Please discard my version of AspNetCore.sln. It was inadvertently pushed to my repository

yogigrantz commented:
Could someone please help me out? I accidentally pushed AspnetCore.sln. Now it is in conflict. I'd appreciate if someone with write access can resolve conflicts by replacing my version with the current master version. Thank you!

BrennanConroy commented:
You should be able to fix it fairly easily:
```
git revert --hard HEAD~2
# redo the parameter naming change
git add <file>
git commit -m "param rename"
git push -f
```

yogigrantz commented:
@BrennanConroy when I typed git revert --hard HEAD~2 it would not take. it gave me the help menu, shown in this screenshot:
![gitCapture](https://user-images.githubusercontent.com/14880717/96660834-78f1b900-12ff-11eb-903b-6173699d3e48.JPG)

BrennanConroy commented:
Whoops, I meant reset. `git reset --hard HEAD~2`

yogigrantz commented:
Brennan: I followed your steps. Not sure if the .sln file was recovered? But could I have another rebuild please? I think the last build broke with #L39

Commit:
Renamed the parm back

Renamed the parm back. Hoping to get a clean build

yogigrantz commented:
@wtgodbe  : is there anything I could do to fix this error ?src/Logging.AzureAppServices/src/AzureAppServicesLoggerFactoryExtensions.cs(39,39): error RS0016: (NETCORE_ENGINEERING_TELEMETRY=Build) Symbol 'AddAzureWebAppDiagnostics' is not part of the declared API.

dougbu commented:
> is there anything I could do to fix this error ?

If you have Visual Studio 16.8 or later installed,
1. `.\restore.cmd` to ensure the .dotnet/ folder is up-to-date
2. `.\startvs.cmd src\Logging.AzureAppServices\src\Microsoft.Extensions.Logging.AzureAppServices.csproj`
3. build the project e.g. *[F6]**
4. click on the RS0016 message in the error window
5. use the light bulb menu or right click the line and add the method to the public API files
6. commit the change and update this branch 😺

yogigrantz commented:
Is VS 16.8 available to download? because when I checked my VS it says 16.7.6 and that's the latest update. When I followed your steps I did not get any errors. I took the screenshots:
![prepOK](https://user-images.githubusercontent.com/14880717/96752373-396cb080-1383-11eb-9472-e05e1eb62b57.JPG)
![buildsOK](https://user-images.githubusercontent.com/14880717/96752385-3ec9fb00-1383-11eb-9568-121347121887.JPG)

dougbu commented:
> Would this work in VS 16.7.6?

@mavasani @sharwell do the public API analyzers and fixers work correctly in 16.7.6❔ I'm surprised @yogigrantz's build was successful.

@yogigrantz it's possible VS thought the project was up-to-date. You may need to rebuild instead of build to get the error seen on the CI.

Another idea would be to use the public preview version of Visual Studio. Easy way to get that is `.\eng\scripts\InstallVisualStudio.ps1 -Channel Preview -Edition Community`.

yogigrantz commented:
@dougbu  : The rebuild works both in 16.7.6 and 16.8 Preview 5.0.  When I downloaded preview 3.1 it gave me version 5.0 btw so I could not get 3.1.

The .\restore.cmd failed on 16.8 preview 5.0. 

Also, when I ran this script: .\eng\scripts\InstallVisualStudio.ps1 -Channel Preview -Edition Community it was pointing to version 16.7.6

dougbu commented:
> The rebuild works both in 16.7.6 and 16.8 Preview 5.0.

I don't understand this and hope @mavasani or @sharwell can help out here.

> When I downloaded preview 3.1 it gave me version 5.0 btw so I could not get 3.1

I didn't mention a specific preview of VS 16.8. Preview 5 is fine.

> it was pointing to version 16.7.6

What pointed to version 16.7.6❔ If you mean the version of VS that startvs.cmd chose to start, That would mean the default is still your previous VS installation.

You can change the association for `*.csproj` files to use the Preview version of VS by default. Or, you can set `$env:VSINSTALLDIR` to `C:\Program Files (x86)\Microsoft Visual Studio\2019\Community_Pre"` before using .\startvs.cmd. The InstallVisualStudio.ps1 command I provided should have installed VS 16.8 there.

mavasani commented:
@dougbu @yogigrantz You are likely hitting https://github.com/dotnet/msbuild/issues/3046, which is applicable to every MSBuild warning, not just analyzer warnings.

yogigrantz commented:
@dougbu : What pointed to version 16.7.6❔  Sorry for the confusion. when I copied your command line, it errored out. So I went ahead and opened InstallVisualStudio.ps1 with powershell and ran it manually. At that point it opened up the Installer option dialog for Enterprise edition, 16.7.6 . After that I changed the value for $Edition and $Channel as shown on the screenshot, but all it does is open the installer dialog with items that I already installed checked out. Is this step for verifying whether I have the correct Visual Studio?
![VSInstaller](https://user-images.githubusercontent.com/14880717/96791095-b3fdf600-13ac-11eb-9d4b-269d3a25f57a.JPG)

yogigrantz commented:
@mavasani : In our case we don't have any warning though. I build with --no-incremental and still, no warning

![nowarning](https://user-images.githubusercontent.com/14880717/96791792-f7a52f80-13ad-11eb-811b-2fb00e4c63af.JPG)

dougbu commented:
@yogigrantz for the questions you asked me: The InstallVisualStudio.ps1 file is intended to run from within PowerShell and we don't have a `.bat` or `.cmd` file anywhere to make it easier to use from `cmd`. When run without `-quiet` or `-passive`, the script does indeed open the installer dialogue. You need to click `Install`, `Modify`, or `Update` to complete the installation or upgrade.

yogigrantz commented:
@dougbu ok I followed thru with update and everything went well. I did another build, it built successfully. 0 Warning 0 Error. There is no changed file in my repository, so there is nothing to commit or push. @mavasani  : when there is no warning and no error, does this situation in  dotnet/msbuild#304 still apply? Could you please advise how I should resolve it?

yogigrantz commented:
Hi, since my repository was too far behind I deleted it and reforked and recloned aspnetcore from the master branch. The current global.json says sdk "version": "6.0.100-alpha.1.20523.3". Should I download this sdk to build the solution, because I have 5.0-rc2 in my machine and it would not build this solution now. If I modify global.json chances are, when I try to merge it would fail the build again. Please advise, thank you!
@dougbu  @mavasani @wtgodbe

dougbu commented:
> Should I download this sdk to build the solution, because I have 5.0-rc2 in my machine and it would not build this solution now.

Run the `./build.cmd` script w/ pretty much any option to install the required SDK and runtimes in the .dotnet/ folder. The `startvs.cmd` ensures Visual Studio uses that folder and not the globally-installed `dotnet`. All command-line builds also ignore the globally-installed `dotnet`.

yogigrantz commented:
Started over from fresh copy of dotnetcore master branch 10/3/2020 -> pull request #27383. I updated / fixed the UnitTest and added the overload in  src/Logging.AzureAppServices/src/PublicAPI.Unshipped.txt  . Now getting only 3 failures
. Could you please let me know whether this failure come from the CI or from the application? Thank you!
@dougbu  @wtgodbe

dougbu commented:
@yogigrantz you need to use Visual Studio to update the src/Logging.AzureAppServices/src/PublicAPI.Unshipped.txt file. It should include the added `AddAzureWebAppDiagnostics` overload.

You may need to remove the .\artifacts\bin\Microsoft.Extensions.Logging.AzureAppServices\ folder and perhaps .\artifacts\obj\Microsoft.Extensions.Logging.AzureAppServices\ to ensure the VS build actually hits the error.

yogigrantz commented:
Transferring to #27383 -> Fresh reclone
